collections/_article/d3d12-shaders-with-a-twist.md

```diff
@@ -0,0 +1,112 @@
+
+### 2. Prevention of NIR level optimization
+
+Mesa's NIR machinery is powerful. It can apply a variety of optimization passes to the NIR representation of a shader. That's awesome, but an issue for us because if we don't inhibit it somehow, we end up again with a "fossilized" shader, only that this time it's obeying some weird sentinel values instead of sensible SC values. Whatever the case, that's bad for our purposes.
+
+The core issues lies in the fact that NIR has a _load constant_ opcode (`nir_intrinsic_load_constant`), which is used to "load" the sentinel value into some virtual register, and that the myriad of optimization passes NIR is run through can optimize it out (if it's possible to pre-compute some ALU operation or however else take a static decision about the fate of that value, for instance.)
+
+The solution was simply to invent a new opcode (`nir_intrinsic_load_constant_non_opt`), which works exactly like the original one, but, by being in the end another one, is unkown to the optimization passes, which have no option but leaving it in place.
+
+### 3. Obtaining patchable DXIL
+
+The last thing we ask Mesa to do for us is generating DXIL from the nicely patched NIR we have at the moment. But our "deliverable" consists also of a table of the bit offsets to patch in the DXIL for each SC id.
+
+I patched the code that emits constants to the DXIL stream so that it recognizes cases of SCs if they match the sentinel value mask. The rest of the bits are the SC id. On each occurrence, the corresponding callback is leveraged to provide the information to the Godot D3D12 driver. (It's a bit more complicated than that, since the offset in the bitcode has later to be adjusted to match where the `DXBC` chunk ends up in the whole DXIL blob.)
+
+### 4. Creating pipelines with patched DXIL
+
+The Direct3D rendering driver is now able to apply whatever specialization constant values are needed to the already compiled shader, which was our initial goal.
+
+Now I have to mention that LLVM bitcode is an akward beast and that it's not trivial to patch a miserable integer due to variable bit-rate encoding. (Speaking of that, the former approach I implemented had a limit on the number of bits that could be patched, which imposed big limitations in the usable range of integers and imposed an elephant-sized epsilon for floating point ones; those have been happily lifted in the current one.)
+
+This is the comparison of the assembly code produced by the same shader, only with different SC values, which allowed in the second case to optimize out a multiplication:
+![Sample ISA code A](/storage/blog/d3d12/isa-a.png)
+![Sample ISA code B](/storage/blog/d3d12/isa-b.png)
```

Comment:
```suggestion
This is the comparison of the assembly code produced by the same shader, only with different SC values, which allowed in the second case to optimize out a multiplication:

![Sample ISA code A](/storage/blog/d3d12/isa-a.png)

![Sample ISA code B](/storage/blog/d3d12/isa-b.png)
```

Reply:
I see the point of the first line break. But in the other case what I actually want is both images side-by-side. They are very thin and will likely fit the screen width everywhere (they second one will just be wrapped into the next line otherwise, won't it?). Just let me know if by doing this I'm missing some rule we have, some layout limitation, or whatever else.

Reply:
We can add custom markup to make them appear side by side on wide enough screens and one under the other on mobile.

Reply:
This is how it ends up looking 
![image](https://user-images.githubusercontent.com/16521339/231867109-dceed9ca-56d1-49f4-bd41-0a7af51ab66b.png)

Reply:
![isa-joined](https://user-images.githubusercontent.com/16521339/231869135-4f6a7901-73cc-4748-b0fd-4ed1d0bc657a.png)

Its probably easiest to just manually join them like this

Reply:
Adding custom markup is easy, and will make sure it works nice on mobile. If you merge them together, it won't work nice on mobile at all. I'll add a suggestion with the markup tomorrow, we should have it in some earlier posts, so it shouldn't take long.

Reply:
I have a big temptation to use a table (like in the good old times). I can use some divs with flex layout, but I have the impression that tables will survive RSS readers better, despite how non-semantic that would be.

Reply:
I didn't realize tables wouldn't work if I want wrapping. I've pushed with some flex-based styling. Would that work?

Reply:
Looks like no

![Screenshot from 2023-04-13 14-41-16](https://user-images.githubusercontent.com/16521339/231889201-f2f029ca-5f6a-4c5a-9f9f-a2579d297d41.png)

Reply:
Thanks for testing, Clay. Could you do one more? If it doesn't work this time, I'll start doing it myself.

Reply:
Note that we don't put full text in our RSS feed :)

Reply:
Maybe you could also center the images?

Reply:
Done!